### PR TITLE
feat(logger): accept arbitrary keyword=value for ephemeral metadata

### LINKED ISSUE #1167: Logger to support arbitrary key=value style arguments in log statements
> Original author: @CeeBeeCee 

**Runtime:** Python

**Is your feature request related to a problem? Please describe**

As per [discussion](https://github.com/awslabs/aws-lambda-powertools-python/discussions/399) originally posted on Python repo, this feature request is related to Logger being able to use key value arguments in log statements e.g. `logger.info("my message", my_key="value", request_id="123d023f0-alk322f")`.

**Describe the solution you'd like**

I'd like to add additional keys on a per log statement basis by passing an arbitrary number of key=value arguments to `.info, .warn, etc.`.

**Describe alternatives you've considered**

1. Use `extra={}` parameter as supported by Logger and in standard logging library
2. Continue to use [structlog](https://github.com/hynek/structlog)

**If you provide guidance, is this something you'd like to contribute?**

**Additional context**

### PR DESCRIPTION
**Pull request by heitorlessa**
**Issue number:** #1167

> This supersedes PR #1505 as (1) Noah is busy with his day-to-day work priorities, and (2) the actual solution requires a full rewrite of the original PR

## Summary

This PR [backports the necessary code implemented in CPython 3.8](https://bugs.python.org/issue33165), monkeypatch and adds additional guards to only use the backported code in older Python versions to not impact performance and future improvements.

### Changes

> Please provide a summary of what's being changed

This change allows customers to pass arbitrary and ephemeral keyword=value arguments in log statements.

It addresses one of the oldest issues and makes it a better UX than `extra` argument. Both can coexist. This is a backport feature from Python 3.8.

**Why did it take us so long?**

The reason that this took us so long was that we inject the `location` key in log, so customers know where a logging statement came from in their code. 

```json
{
    "level": "WARNING",
    "location": "warning:465",
    "message": "Testing",
    "timestamp": "2022-10-26 16:22:27,543+0200",
    "service": "payment",
    "request_id": "0c7ff166-3be1-4795-8f0c-b48ca5ce5a95"
}
```

When using a class like `Logger`, we need to override all log statement methods. When doing so, it moves the stack frame location, where the `location` key now points to where the statement was called inside `Logger`, **instead** of the customer log statement location.

```python
def warning(msg, *args, **kwargs):...
def info(msg, *args, **kwargs):
    return self._logger.info(msg, *args, **kwargs)  # <--- location would always show up here
....
```

With this change, the `location` continues to be precise as we move the stack frame location.

```json
{
    "level": "WARNING",
    "location": "<module>:7",
    "message": "Testing",
    "timestamp": "2022-10-26 15:54:34,945+0200",
    "service": "payment",
    "request_id": "c344d8ca-2697-42b1-ad67-2bed036c550d",
}
```

### User experience

> Please share what the user experience looks like before and after this change

**BEFORE**

```python
from aws_lambda_powertools import Logger

from uuid import uuid4

logger = Logger(service="payment")

logger.warning("Testing", extra={"request_id": f"{uuid4()}"})
```

**AFTER**

```python
from aws_lambda_powertools import Logger

from uuid import uuid4

logger = Logger(service="payment")

logger.warning("Testing", request_id=f"{uuid4()}")
```

## Checklist

If your change doesn't seem to apply, please leave them unchecked.

* [x] [Meet tenets criteria](https://awslabs.github.io/aws-lambda-powertools-python/#tenets)
* [x] I have performed a self-review of this change
* [x] Changes have been tested
* [x] Changes are documented
* [x] PR title follows [conventional commit semantics](https://github.com/awslabs/aws-lambda-powertools-python/blob/develop/.github/semantic.yml)

<details>
<summary>Is this a breaking change?</summary>

**RFC issue number**:

Checklist:

* [ ] Migration process documented
* [ ] Implement warnings (if it can live side by side)

</details>

## Acknowledgment

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**Disclaimer**: We value your time and bandwidth. As such, any pull requests created on non-triaged issues might not be successful.

### CODE REVIEW

**Comment by heitorlessa** [comment]
Waiting for CI to complete to ensure it works in a fresh Python env, then will refresh Docs.

**Comment by heitorlessa** [comment]
Need to add co-authorship for Noah before merging (@OGoodness). I'll work on docs now as we've finished peer review due to the complexity of this change.

**Comment by heitorlessa** [comment]
Docs now in!

image

**Review by leandrodamascena — APPROVED** [review]
Amazing work @heitorlessa! Thanks for sharing the ideas you had for changing the code this way and make it simple.

image

**Pull request merged**
